# **CMPT 353: Project Report**

## **1. Abstract**

## **2. Introduction**

## **3. Data Gathering and Retrieval**

## **4. Data Analysis**

### **4.1 Represention of Lyrics Features**

To better compute a score for lyrics pertaining to a specific song in a playlist as a part of an entire playlist' lyrics dataset, we introduce the Term Frequency-Inverse Document Frequency (or TFIDF for short). The TFIDF score depicts the number of occurences of a word of a lyric in relation to an entire list of words throughout a lyrics document [1].

We begin by determining the TF score for each word within a particular lyric using the formula below:

$$ 
TF_{i,j} = \frac{|number\space \,of\space \,occurences\space \,of\space \,word\space \,i\space \,within\space \,a\space \,lyric\space \,j|}{|lyric\space \,j|}\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\space\,[1]
$$

Next, we calculate the inverse document frequenccy score for each lyric within the document

$$ 
IDF_{i} = \frac{|total\space\,number\space \,of\space \,lyrics\space \,in\space \,dataset\space \,j|}{|number\space\,of\space \,lyrics\space \,containing\space \,word\space\,i|}\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace[1]$$

The resulting TFIDF score from the computations becomes:

$\enspace\,TFIDF_{i,j} = TF_{i,j}\space\,\times\,\space\,IDF_{i}\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,\enspace\,[1]$

Upon performing the computation in the lyrics_tfidf.py file we store the output into a csv file composed of the tfidf scores and the words in all the lyrics.
This is accomplished by running the following commands consecutively:

##### i. python3 lyrics.py playlist.csv

##### ii. python3 lyrics_tfidf.py playlist.csv

A clip of the output as can be seen in the Lyrics_TFIDF_Score file is below:

In [19]:
import pandas as pd
df = pd.read_csv('Lyrics_TFIDF_Score')
df

Words  TFIDF_Score_Per_Word
0     everlasting             53.730725
1           choke             53.716703
2          wisdom             53.618552
3       spineless             53.604530
4           chill             53.562466
...           ...                   ...
2434          get              0.070501
2435         want              0.031521
2436         tell              0.013163
2437           yo              0.004385
2438       hahaha              0.002909

[2439 rows x 2 columns]

## 5. Results

### 5.1. Statistical Tests

In the subsequent section we would make use statistical tests to determine whether there is any form of correlation between words with the highest TFIDF scores and the popularity of songs in a playlist.
We firstly begin by computing the average tfidf scores of words in a playlist and testing it alongside the popularity of each song in a playlist to observe whether a correlation might exist using the linear regression model.

#### 5.1.1. Linear Regression Model

Incorporating linear regression using the *stats.linregress* python module and evaluating the average TFIDF scores across all playlists elicits the output below:

In [20]:
run -i 'stat_tests.py playlist.csv'

The slope of the regression model is:  -0.07466477444163927
The intercept of the regression model is:  73.98577779631047
The p-value of the regression model is:  0.6580528892070687
The regression coefficient is:  -0.06415155325179604 

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.492
Model:                            OLS   Adj. R-squared (uncentered):              0.481
Method:                 Least Squares   F-statistic:                              47.37
Date:                Tue, 12 Apr 2022   Prob (F-statistic):                    1.01e-08
Time:                        01:54:30   Log-Likelihood:                         -195.16
No. Observations:                  50   AIC:                                      392.3
Df Residuals:                      49   BIC:                                      394.2
Df Model:                           1                       

Hence, from the output above, although the p-value of the linear model is > 0.05 there is only a weak correlation of approximately -0.064. This entails a conclusion that the average TFIDF score of a lyric from a song in a playlist is nearly  independent of the popularity score of the song.

### 5.2 Visualization

#### 5.2.1. $\space$Formulating word frequency based on each word's TFIDF Score

To conceptualize the relevancy of the TFIDF scores of each word, we would integrate the use of a wordcloud.
A module containing a wordcloud generator is accessible in python by entering using the command below:

##### pip install wordcloud

The wordcloud works by filtering through each word in a lyric while storing each new occurence in a list. After each word is appended onto a list, we feed the list to a WordCloud imported class. The final .png file is computed and stored in the current directory. This is done by running the command below: 
<div>
    <span>
        <h5> python3 wordcloud.py playlist.csv <h5>
    </span>
    <img src="wordcloud.png" width="500"/>
    <span>
        <h5>
            <i> 
                <center>
                    Fig. 1. WordCloud Image depicting the frequency of the words in a playlist 
                </center>
            <i>
        </h5>
    </span>
</div>

## 6. Limitations

## 7. Conclusion

## 8. References

[1]. https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.650.7424&rep=rep1&type=pdf